### PR TITLE
Rocm/fused short seq attention

### PR DESCRIPTION
**Pull request by MohamedSayedFathy**
## Purpose

Optimize PagedAttention v2 decode performance on ROCm for short sequences (≤256 tokens) by eliminating the reduce kernel when only a single partition is needed.

When `max_num_partitions == 1`, the QKV kernel writes directly to `final_out` instead of `tmp_out`, skipping:
- The `tmp_out` intermediate buffer write (~8 KB per seq/layer)
- The `tmp_out` read-back in the reduce kernel (~8 KB per seq/layer)
- The `exp_sums`/`max_logits` metadata writes (~256 bytes per seq/layer)
- The reduce kernel launch entirely (~10μs overhead)

This saves ~16.5 KB of HBM traffic per sequence per layer and eliminates one kernel launch.

### Key design choices:
- Uses `if constexpr (SINGLE_PARTITION)` for zero-cost abstraction — dead code eliminated at compile time
- Enabled by default, togglable via `VLLM_ROCM_FUSED_SHORT_SEQ=0` for A/B benchmarking
- Applied to both mfma4 (GQA ≤ 4) and mfma16 (GQA 5-16) kernels on MI300X
- No changes to the multi-partition path — long sequences are unaffected

## Test Plan

Verified on AMD MI300X by running `benchmark_paged_attention.py` with both `VLLM_ROCM_FUSED_SHORT_SEQ=1` (optimized) and `VLLM_ROCM_FUSED_SHORT_SEQ=0` (baseline). Both paths produce correct results with matching outputs.

```bash
# Optimized path (default)
python benchmarks/kernels/benchmark_paged_attention.py \
  --version v2 --custom-paged-attn --seq-len 128 --batch-size 8

# Baseline path (disabled)
VLLM_ROCM_FUSED_SHORT_SEQ=0 python benchmarks/kernels/benchmark_paged_attention.py \
  --version v2 --custom-paged-attn --seq-len 128 --batch-size 8

### CODE REVIEW

**Comment by MohamedSayedFathy** [comment]
Hi @tjtanaa & @gshtras, this is my first contribution to vllm, and I believe it's blocked by the new-contributor CI gate (pre-run-check requires the ready label or 4+ merged PRs). All pre-commit fixes have been applied, and the branch is up to date with main. Could a maintainer add the ready label so CI can run and the PR can be reviewed? Happy to address any feedback.
